In [1]:
import pandas as pd
import numpy as np

# Data import and cleaning

## Exploring WVS time series 5.0

In [2]:
df = pd.read_csv('../WVS_time_series/WVS_Time_Series_1981-2022_v5_0.csv')

In [3]:
df

version                        doi  S001  S002VS  S003  \
0       5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3     8   
1       5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3     8   
2       5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3     8   
3       5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3     8   
4       5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3     8   
...                    ...                        ...   ...     ...   ...   
443483  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       7   716   
443484  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       7   716   
443485  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       7   716   
443486  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       7   716   
443487  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       7   716   

       COUNTRY_ALPHA  COW_NUM COW_ALPHA  S004       S006  ...     Y023A  \
0                ALB      339       ALB     1          1  ...  0.000000   
1                ALB      339       ALB     1          2  ...  0.111111   
2                ALB      339       ALB     1          3  ...  0.000000   
3                ALB      339       ALB     1          4  ...  0.000000   
4                ALB      339       ALB     1          5  ...  0.000000   
...              ...      ...       ...   ...        ...  ...       ...   
443483           ZWE      552       ZIM    -4  716071211  ...  0.000000   
443484           ZWE      552       ZIM    -4  716071212  ...  0.000000   
443485           ZWE      552       ZIM    -4  716071213  ...  0.000000   
443486           ZWE      552       ZIM    -4  716071214  ...  0.000000   
443487           ZWE      552       ZIM    -4  716071215  ...  0.000000   

           Y023B     Y023C  Y024A  Y024B  Y024C  survself  tradrat5   TradAgg  \
0       0.444444  0.444444   0.33    0.0  0.165 -1.011276  0.160847  0.158964   
1       0.444444  0.444444   0.33    0.0  0.165 -0.688773  0.409283  0.558946   
2       0.444444  0.444444   0.33    0.5  0.415       NaN       NaN       NaN   
3       0.333333  0.333333   0.33    0.0  0.165 -1.319863  0.733956  1.081669   
4       0.333333  0.333333   0.00    0.5  0.250       NaN       NaN       NaN   
...          ...       ...    ...    ...    ...       ...       ...       ...   
443483  0.000000  0.000000   0.66    0.5  0.580 -0.738923 -0.774148 -1.346378   
443484  0.000000  0.777778   1.00    0.0  0.500 -0.445039  0.282275  0.354463   
443485  0.000000  0.000000   0.33    0.0  0.165 -0.925464 -0.436454 -0.802691   
443486  0.000000  0.000000   0.00    0.0  0.000 -2.398929  1.218936  1.862486   
443487  0.000000  0.000000   0.00    0.0  0.000 -1.633334  0.249829  0.302224   

        SurvSAgg  
0      -1.792410  
1      -1.208679  
2            NaN  
3      -2.350952  
4            NaN  
...          ...  
443483 -1.299450  
443484 -0.767520  
443485 -1.637090  
443486 -4.304062  
443487 -2.918335  

[443488 rows x 1046 columns]

In [4]:
df[['S003', 'COUNTRY_ALPHA']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443488 entries, 0 to 443487
Data columns (total 2 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   S003           443488 non-null  int64 
 1   COUNTRY_ALPHA  443488 non-null  object
dtypes: int64(1), object(1)
memory usage: 6.8+ MB


All of the country values are complete - so we can filter safely only for the United States values.

In [5]:
df['S002VS'].value_counts()

S002VS
7    97220
6    89565
5    83975
3    77818
4    60045
2    24558
1    10307
Name: count, dtype: int64

In [6]:
df_first2waves = df[(df['S002VS'] == 1) | (df['S002VS'] == 2)]

for column in ['S002VS', 'S003', 'COUNTRY_ALPHA']:
    print(df_first2waves[column].value_counts())

S002VS
2    24558
1    10307
Name: count, dtype: int64
S003
710    4332
484    3368
356    2500
410    2221
392    2215
32     2007
643    1961
76     1782
724    1510
152    1500
348    1464
756    1400
36     1228
792    1030
112    1015
246    1003
566    1001
156    1000
616     938
203     924
703     466
Name: count, dtype: int64
COUNTRY_ALPHA
ZAF    4332
MEX    3368
IND    2500
KOR    2221
JPN    2215
ARG    2007
RUS    1961
BRA    1782
ESP    1510
CHL    1500
HUN    1464
CHE    1400
AUS    1228
TUR    1030
BLR    1015
FIN    1003
NGA    1001
CHN    1000
POL     938
CZE     924
SVK     466
Name: count, dtype: int64


In [7]:
df_first2waves[['S002VS', 'S003', 'COUNTRY_ALPHA']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 34865 entries, 4006 to 425816
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   S002VS         34865 non-null  int64 
 1   S003           34865 non-null  int64 
 2   COUNTRY_ALPHA  34865 non-null  object
dtypes: int64(2), object(1)
memory usage: 1.1+ MB


All the waves are included! But why is there no information for the United States, when theoretically there should also be questions for waves 1 and 2?

In [8]:
df_us = df[df['COUNTRY_ALPHA'] == 'USA']

df_us

version                        doi  S001  S002VS  S003  \
401631  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3   840   
401632  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3   840   
401633  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3   840   
401634  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3   840   
401635  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3   840   
...                    ...                        ...   ...     ...   ...   
410445  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       7   840   
410446  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       7   840   
410447  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       7   840   
410448  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       7   840   
410449  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       7   840   

       COUNTRY_ALPHA  COW_NUM COW_ALPHA  S004       S006  ...     Y023A  \
401631           USA        2       USA     1          1  ...  0.000000   
401632           USA        2       USA     1          2  ...  0.555556   
401633           USA        2       USA     1          3  ...  1.000000   
401634           USA        2       USA     1          4  ...  0.000000   
401635           USA        2       USA     1          5  ...  0.000000   
...              ...      ...       ...   ...        ...  ...       ...   
410445           USA        2       USA    -4  840072715  ...  0.666667   
410446           USA        2       USA    -4  840071538  ...  0.888889   
410447           USA        2       USA    -4  840071446  ...  0.444444   
410448           USA        2       USA    -4  840071167  ...  0.555556   
410449           USA        2       USA    -4  840072217  ...  0.555556   

           Y023B     Y023C  Y024A  Y024B  Y024C  survself  tradrat5   TradAgg  \
401631  0.000000  0.444444   0.00    0.0  0.000  0.557763 -1.243387 -2.101853   
401632  0.666667  0.444444   0.66    1.0  0.830  2.080791 -0.781275 -1.357852   
401633  1.000000  1.000000   1.00    0.0  0.500       NaN       NaN       NaN   
401634  0.000000  0.000000   0.33    0.5  0.415  0.956679 -1.568910 -2.625945   
401635  1.000000  1.000000   1.00    0.0  0.500  1.020697 -0.169883 -0.373511   
...          ...       ...    ...    ...    ...       ...       ...       ...   
410445  0.555556  0.555556   0.00    0.0  0.000       NaN       NaN       NaN   
410446  0.777778  0.777778   1.00    1.0  1.000  0.312523  2.282574  3.574944   
410447  0.555556  0.555556   1.00    1.0  1.000  0.999833 -0.738978 -1.289755   
410448  0.000000  0.777778    NaN    NaN    NaN       NaN       NaN       NaN   
410449  0.000000  0.777778   0.00    0.5  0.250  1.049813 -1.685598 -2.813813   

        SurvSAgg  
401631  1.047551  
401632  3.804231  
401633       NaN  
401634  1.769589  
401635  1.885462  
...          ...  
410445       NaN  
410446  0.603666  
410447  1.847697  
410448       NaN  
410449  1.938161  

[8819 rows x 1046 columns]

In [9]:
df_us['S002VS'].value_counts()

S002VS
7    2596
6    2232
3    1542
5    1249
4    1200
Name: count, dtype: int64

Note how there is no data for the first 2 waves. This is solved below by merging the latest version of the dataset (above) with a previous version, which includes the data for the US.

## Missing US data for the first 2 waves

In [12]:
df2020 = pd.read_csv('../WVS_time_series/WVS_TimeSeries_1981_2020_ascii_v2_0.csv')

df2020

S001  S002VS  S003 COUNTRY_ALPHA  COW_NUM COW_ALPHA  S006        S007  \
0          2       1    32           ARG      160       ARG     1   320120001   
1          2       1    32           ARG      160       ARG     2   320120002   
2          2       1    32           ARG      160       ARG     3   320120003   
3          2       1    32           ARG      160       ARG     4   320120004   
4          2       1    32           ARG      160       ARG     5   320120005   
...      ...     ...   ...           ...      ...       ...   ...         ...   
432477     2       7   716           ZWE      552       ZIM  1211  7160721211   
432478     2       7   716           ZWE      552       ZIM  1212  7160721212   
432479     2       7   716           ZWE      552       ZIM  1213  7160721213   
432480     2       7   716           ZWE      552       ZIM  1214  7160721214   
432481     2       7   716           ZWE      552       ZIM  1215  7160721215   

        S008  S010  ...  Y022B  Y022C     Y023A     Y023B     Y023C  Y024A  \
0         -4  -4.0  ...    NaN    NaN  0.000000  0.000000  0.000000   0.00   
1         -4  -4.0  ...    NaN    NaN  0.000000  0.000000  0.000000   0.00   
2         -4  -4.0  ...    NaN    NaN  0.111111  0.333333  0.555556   0.00   
3         -4  -4.0  ...    NaN    NaN  0.333333  0.555556  0.666667   1.00   
4         -4  -4.0  ...    NaN    NaN  0.444444  0.222222  0.333333   0.00   
...      ...   ...  ...    ...    ...       ...       ...       ...    ...   
432477    50  77.0  ...   0.00   0.66  0.000000  0.000000  0.000000   0.66   
432478    50  80.0  ...   0.00   1.00  0.000000  0.000000  0.777778   1.00   
432479    60  59.0  ...   0.66   0.66  0.000000  0.000000  0.000000   0.33   
432480    60  83.0  ...   0.66   0.66  0.000000  0.000000  0.000000   0.00   
432481    60  86.0  ...   0.66   0.66  0.000000  0.000000  0.000000   0.00   

        Y024B  Y024C   TradAgg  SurvSAgg  
0         NaN    NaN -1.358583 -1.124587  
1         NaN    NaN -0.142223 -2.673256  
2         NaN    NaN  0.084509 -2.033451  
3         NaN    NaN  3.529956 -0.442163  
4         NaN    NaN  1.913126 -1.911873  
...       ...    ...       ...       ...  
432477    0.5  0.580 -1.365047 -1.280893  
432478    0.0  0.500  0.315010 -0.756891  
432479    0.0  0.165 -0.844314 -1.613268  
432480    0.0  0.000  1.775634 -4.353077  
432481    0.0  0.000  0.236233 -2.928095  

[432482 rows x 1072 columns]

In [13]:
df_first2waves = df2020[(df2020['S002VS'] == 1) | (df2020['S002VS'] == 2)]

for column in ['S002VS', 'S003', 'COUNTRY_ALPHA']:
    print(df_first2waves[column].value_counts())

S002VS
2    28127
1    14840
Name: count, dtype: int64
S003
710    4332
840    4164
484    3368
124    2984
356    2500
410    2221
392    2215
32     2007
643    1961
76     1782
724    1510
152    1500
348    1464
756    1400
36     1228
792    1030
112    1015
246    1003
566    1001
156    1000
752     954
616     938
203     924
703     466
Name: count, dtype: int64
COUNTRY_ALPHA
ZAF    4332
USA    4164
MEX    3368
CAN    2984
IND    2500
KOR    2221
JPN    2215
ARG    2007
RUS    1961
BRA    1782
ESP    1510
CHL    1500
HUN    1464
CHE    1400
AUS    1228
TUR    1030
BLR    1015
FIN    1003
NGA    1001
CHN    1000
SWE     954
POL     938
CZE     924
SVK     466
Name: count, dtype: int64


This includes data for the US! Now we will only keep those observations of the US, to concatenate afterwards the datasets.

In [14]:
df_us_missing = df_first2waves[df_first2waves['COUNTRY_ALPHA'] == 'USA']

df_us_missing

S001  S002VS  S003 COUNTRY_ALPHA  COW_NUM COW_ALPHA  S006        S007  \
12515     1       1   840           USA        2       USA     1  8400120001   
12516     1       1   840           USA        2       USA     2  8400120002   
12517     1       1   840           USA        2       USA     3  8400120003   
12518     1       1   840           USA        2       USA     4  8400120004   
12519     1       1   840           USA        2       USA     5  8400120005   
...     ...     ...   ...           ...      ...       ...   ...         ...   
42962     1       2   840           USA        2       USA  1835  8400221835   
42963     1       2   840           USA        2       USA  1836  8400221836   
42964     1       2   840           USA        2       USA  1837  8400221837   
42965     1       2   840           USA        2       USA  1838  8400221838   
42966     1       2   840           USA        2       USA  1839  8400221839   

       S008  S010  ...  Y022B  Y022C     Y023A     Y023B     Y023C  Y024A  \
12515    -4  -4.0  ...    NaN    NaN  0.555556  0.444444  0.444444    NaN   
12516    -4  -4.0  ...    NaN    NaN  0.444444  0.444444  0.444444    NaN   
12517    -4  -4.0  ...    NaN    NaN       NaN       NaN       NaN    NaN   
12518    -4  -4.0  ...    NaN    NaN  0.000000  0.000000  0.000000    NaN   
12519    -4  -4.0  ...    NaN    NaN  0.000000  0.444444  0.666667    NaN   
...     ...   ...  ...    ...    ...       ...       ...       ...    ...   
42962    -4  -4.0  ...    NaN    NaN  0.000000  0.555556  0.555556   0.66   
42963    -4  -4.0  ...    NaN    NaN  0.000000  0.000000  0.000000   0.33   
42964    -4  -4.0  ...    NaN    NaN  0.444444  0.666667  0.666667   1.00   
42965    -4  -4.0  ...    NaN    NaN  0.000000  0.555556  0.555556    NaN   
42966    -4  -4.0  ...    NaN    NaN  0.444444  1.000000  1.000000   0.33   

       Y024B  Y024C   TradAgg  SurvSAgg  
12515    NaN    NaN       NaN       NaN  
12516    NaN    NaN       NaN       NaN  
12517    NaN    NaN       NaN       NaN  
12518    NaN    NaN       NaN       NaN  
12519    NaN    NaN       NaN       NaN  
...      ...    ...       ...       ...  
42962    NaN    NaN       NaN       NaN  
42963    0.5  0.415       NaN       NaN  
42964    0.5  0.750 -0.229636  4.338274  
42965    0.0    NaN       NaN       NaN  
42966    1.0  0.665  1.751926  2.976972  

[4164 rows x 1072 columns]

Check the columns that are different in the 5.0 dataset and the 2.0 dataset.

In [26]:
diff_columns = set(df_us.columns).symmetric_difference(df_us_missing.columns)
print(len(diff_columns))
print(sorted(diff_columns))

64
['A007', 'A080_01', 'A080_02', 'C001_01', 'C002_01', 'D043_01', 'D066_01', 'E069_16', 'E069_18A', 'E069_67', 'E147', 'E154', 'E155', 'E156', 'E157', 'E158', 'E159', 'E160', 'E161', 'E162', 'E179_WVS7LOC', 'E181_EVS5', 'F025', 'F025_01old', 'F025_WVS', 'F025old', 'F030', 'F144_01', 'G022S', 'G033', 'G034', 'G035', 'G036', 'G038', 'G040', 'G041', 'G043', 'G061', 'MODE', 'S004', 'S016', 'S019', 'W001', 'W004', 'W005_2', 'W007', 'W008', 'W009', 'W010', 'W011', 'X002_03', 'X003R', 'X003R2', 'X006_02', 'X007_02', 'X025CS', 'X028_01', 'X032R', 'X034', 'X046', 'doi', 'survself', 'tradrat5', 'version']


Maybe all of these different columns should be dropped. For now, though, we will concatenate the two datasets.

## Merging the data

In [16]:
df_us_concat = pd.concat([df_us, df_us_missing])

df_us_concat

version                        doi  S001  S002VS  S003  \
401631  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3   840   
401632  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3   840   
401633  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3   840   
401634  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3   840   
401635  5-0-0 (2024-04-30)  doi.org/10.14281/18241.25     2       3   840   
...                    ...                        ...   ...     ...   ...   
42962                  NaN                        NaN     1       2   840   
42963                  NaN                        NaN     1       2   840   
42964                  NaN                        NaN     1       2   840   
42965                  NaN                        NaN     1       2   840   
42966                  NaN                        NaN     1       2   840   

       COUNTRY_ALPHA  COW_NUM COW_ALPHA  S004  S006  ...  W010  W011  X002_03  \
401631           USA        2       USA   1.0     1  ...   NaN   NaN      NaN   
401632           USA        2       USA   1.0     2  ...   NaN   NaN      NaN   
401633           USA        2       USA   1.0     3  ...   NaN   NaN      NaN   
401634           USA        2       USA   1.0     4  ...   NaN   NaN      NaN   
401635           USA        2       USA   1.0     5  ...   NaN   NaN      NaN   
...              ...      ...       ...   ...   ...  ...   ...   ...      ...   
42962            USA        2       USA   NaN  1835  ...  -4.0  -4.0     -4.0   
42963            USA        2       USA   NaN  1836  ...  -4.0  -4.0     -4.0   
42964            USA        2       USA   NaN  1837  ...  -4.0  -4.0     -4.0   
42965            USA        2       USA   NaN  1838  ...  -4.0  -4.0     -4.0   
42966            USA        2       USA   NaN  1839  ...  -4.0  -4.0     -4.0   

        X006_02  X007_02  X025CS  X028_01  X032R  X034  X046  
401631      NaN      NaN     NaN      NaN    NaN   NaN   NaN  
401632      NaN      NaN     NaN      NaN    NaN   NaN   NaN  
401633      NaN      NaN     NaN      NaN    NaN   NaN   NaN  
401634      NaN      NaN     NaN      NaN    NaN   NaN   NaN  
401635      NaN      NaN     NaN      NaN    NaN   NaN   NaN  
...         ...      ...     ...      ...    ...   ...   ...  
42962      -4.0     -4.0    -4.0     -4.0   -4.0  -4.0  -2.0  
42963      -4.0     -4.0    -4.0     -4.0   -4.0  -4.0   4.0  
42964      -4.0     -4.0    -4.0     -4.0   -4.0  -4.0   1.0  
42965      -4.0     -4.0    -4.0     -4.0   -4.0  -4.0   3.0  
42966      -4.0     -4.0    -4.0     -4.0   -4.0  -4.0   1.0  

[12983 rows x 1091 columns]

Now, all of the waves are complete for the US!

## Looking for relevant, general questions that can be linked to society's concerns

In [20]:
df_us_concat['D072'].value_counts(normalize=True)

D072
-4    1.0
Name: proportion, dtype: float64

Useful variables:
- `S002VS`: survey wave. 1 1981-1984; 2 1989-1993; 3 1994-1998; 4 1999-2004; 5 2005-2009; 6 2010-2014; 7 2017-2022
- `S003`: country code ISO 3166 (840 for the US).
- `S007`: unified respondent number. Therefore, **each row in the dataset is a different respondent!**
- `S012`: Date of interview YYYYMMDD.
- Weight variables that should be used for aggregating the data by wave (read more about it [here](https://www.worldvaluessurvey.org/WVSContents.jsp?CMSID=WEIGHT&CMSID=WEIGHT)):
  - `S017`: Weight assigned to the respondent. Weight by country.
  - `S018`: Equilibrated weight-1000. According to the reference above, we should use this to consistently aggregate the data per wave and year! In summary, "Each respondent has a weight to ensure national representativeness. Always calculate weighted means or proportions so that your country-level aggregates reflect the population accurately."
- `S020`: Year of survey YYYY.

Useful questions (from page 224-414 of the "Variables Report V5.0):
- **A (society and personal view)**: A001-A006 (importance in life of various topics), A008 (feeling of happiness), A013 (feeling of loneliness), A046-A049 (abortion), A062 (politics discussion with friends), A124_01-61 (preferences, for neighbors), A170 (overall satisfaction with life), A200-A202 (position of people of different ages in society), A208 (respect towards older people in society).
- **B (environment)**: B001-B004, B007 (can be used as proxies for the willingness to take action to tackle climate change and have better relation with the environment), B021-B023 (concern about climate change, biodiversity loss and pollution).
- **C (economy and employment)**: C006 (satisfaction with financial situation of HH), C033 (job satisfaction).
- **D (family)**: D002 (satisfaction with home life), D024 (sexual freedom), D057-60, D063, D068-69, D072, D077 (sexism).
- **E (country and politics)**: E001-010 (preferences towards the country's objectives), E033 (self-positioning in political scale, left-right), E034 (attitude towards society - change/defended), E035 (income inequality), E037 (government paternalism), E041 (wealth inequality), E046 (view towards progression), E057 (economic system needing fundamental changes), E059 (preferences towards general individual freedom), E069_1-67 (confidence towards various organizations and institutions, such as the police, the government, the press, etc.), E104 (approval towards ecology movement), E106 (approval towards disarmament movement), E107 (approval towards human rights movement), E108 (approval towards women's movement), E111 (rate of the political system of the governing country), E112 (rate of the political system as it was before), E113 (rate of political system in future, 10 years), E143 (immigrant policy), E196 (extent of political corruption), E205 (political parties serving the social and political needs of people), E268 (scale of corruption).
- **F (religion, life, death and freedoms)**. F118 (how justifiable homosexuality is), F120 (how justifiable abortion is).
- **G (national identity and immigration)**: G005/G017 (birthplace of surveyed), G006/G021 (proudness of nationality), G018 (year when surveyed came to the country), G032 (impact of ethnic diversity on country), G052 (impact of immigration on development).
- **H (perceived security, other concerns, and government interventionism)**: H001/H008_02 (overall perceived security), H002_02 (frequency of alcohol consumption in neighborhood), H002_05 (frequency of drug sale in neighborhood), H006_03 (concern about war involving the respondent's country).
- **I (science)**.
- **V (demographical characteristics of respondent's parents)**.
- **W (demographical characteristics of respondent's partner/spouse)**.
- **X (demographical characteristics of respondent)**: X001 (sex), X002 (YOB), X002_02A (country of birth), X003 (age), X007 (marital status), X011 (number of children), X025, X025A_01 (highest educational level attained), X028 (employment status).
- **Y (coded indexes of other questions)**: these indexes quantify some aspects of the questions made above. Y011B (nationalism - inverse national pride), Y012A (importance of religion), Y023A (homosexuality acceptance).

After page 224, reference of codes and others.

From the interesting questions above, which are not available for all the waves (after concatenating the data for the 1st and 2nd waves and the data for the 3rd to 7th waves)?

- Complete unmatching columns (checked above): ['A007', 'A080_01', 'A080_02', 'C001_01', 'C002_01', 'D043_01', 'D066_01', 'E069_16', 'E069_18A', 'E069_67', 'E147', 'E154', 'E155', 'E156', 'E157', 'E158', 'E159', 'E160', 'E161', 'E162', 'E179_WVS7LOC', 'E181_EVS5', 'F025', 'F025_01old', 'F025_WVS', 'F025old', 'F030', 'F144_01', 'G022S', 'G033', 'G034', 'G035', 'G036', 'G038', 'G040', 'G041', 'G043', 'G061', 'MODE', 'S004', 'S016', 'S019', 'W001', 'W004', 'W005_2', 'W007', 'W008', 'W009', 'W010', 'W011', 'X002_03', 'X003R', 'X003R2', 'X006_02', 'X007_02', 'X025CS', 'X028_01', 'X032R', 'X034', 'X046', 'doi', 'survself', 'tradrat5', 'version']
- Therefore, from the interesting questions above, we don't have complete data for: E069_16, E069_18A and E069_67 (not relevant).

In [35]:
columns_interest = ['A170', 'B007', 'B021', 'A005', 'C006', 'C033', 'A001',
                    'D002', 'D024', 'F118', 'F120', 'D057', 'D058', 'D059',
                    'D063', 'D072', 'A004', 'E010', 'E046', 'E059', 'E069_06',
                    'E069_07', 'E069_12', 'E069_17',
                    'E069_11', 'E111', 'G006', 'Y011B', 'E106', 'H006_03',
                    'E035', 'A006', 'E196', 'E268', 'H001', 'H008_02', 'H002_02',
                    'H002_05', 'A009', 'E143', 'G032', 'G052', 'H002_04']

for column in columns_interest:
    print(df_us_concat[column].value_counts(normalize = True))

A170
 8     0.267735
 9     0.188631
 7     0.173689
 10    0.131711
 5     0.078795
 6     0.078718
 4     0.034430
 3     0.022337
 2     0.010706
 1     0.008704
-2     0.003312
-1     0.000770
-5     0.000462
Name: proportion, dtype: float64
B007
-4    0.858353
 3    0.072094
 2    0.032119
 4    0.024571
 1    0.006547
-1    0.005546
-2    0.000770
Name: proportion, dtype: float64
B021
-4    0.903797
 1    0.043827
 2    0.031888
 3    0.012709
 4    0.005238
-5    0.001849
-2    0.000693
Name: proportion, dtype: float64
A005
 1    0.362474
 2    0.312332
-4    0.179080
 3    0.086729
 4    0.052376
-2    0.004313
-1    0.002696
Name: proportion, dtype: float64
C006
 8     0.180698
 7     0.161904
 5     0.122468
 6     0.116768
 10    0.100439
 9     0.099900
 4     0.072942
 3     0.054533
 1     0.052607
 2     0.030578
-2     0.005161
-5     0.001078
-1     0.000924
Name: proportion, dtype: float64
C033
-4     0.858353
-3     0.055534
 8     0.021182
 9     0.017484
 10    0.0

In [48]:
df_us_concat_subset = df_us_concat[['S002VS', 'S003', 'S007', 'S017',
                                    'S018', 'S020', 'C006', 'E069_06', 'E069_07',
                                    'E069_17', 'G006', 'A009']]

df_us_concat_subset.head()

S002VS  S003        S007     S017      S018  S020  C006  E069_06  \
401631       3   840  8400320001  0.50021  0.648508  1995     9        2   
401632       3   840  8400320002  0.87037  0.648508  1995     8        1   
401633       3   840  8400320003  1.42561  0.648508  1995     2        3   
401634       3   840  8400320004  0.88037  0.648508  1995     9        2   
401635       3   840  8400320005  0.65028  0.648508  1995     2        2   

        E069_07  E069_17  G006  A009  
401631        2        2     1     2  
401632        3        3     1     1  
401633        3        3     2     3  
401634        2        2     1     1  
401635        3        3     2     1

Viewing the missing/uninteresting values per survey wave:  

In [49]:
columns_of_interest = ['C006', 'E069_06', 'E069_07', 'E069_17', 'G006', 'A009']

for column in columns_of_interest:
    print(pd.crosstab(df_us_concat_subset[column], df_us_concat_subset['S002VS'], 
                      normalize = 'columns', margins = True))

S002VS         1         2         3         4         5         6         7  \
C006                                                                           
-5      0.000000  0.000000  0.000000  0.000000  0.011209  0.000000  0.000000   
-2      0.009462  0.002175  0.000000  0.000000  0.001601  0.007168  0.008860   
-1      0.000000  0.002719  0.003891  0.000833  0.000000  0.000000  0.000000   
1       0.053333  0.044046  0.054475  0.050833  0.057646  0.049283  0.058166   
2       0.032688  0.022295  0.023995  0.025000  0.036829  0.034050  0.035054   
3       0.056344  0.035345  0.054475  0.045833  0.061649  0.062276  0.060478   
4       0.056774  0.056009  0.063554  0.074167  0.088871  0.082885  0.088213   
5       0.116129  0.108755  0.129702  0.105000  0.140112  0.117384  0.137519   
6       0.101075  0.119630  0.105058  0.126667  0.132106  0.108871  0.130586   
7       0.149677  0.142469  0.166667  0.188333  0.180144  0.166667  0.158706   
8       0.175914  0.198477  0.167315  0.

There is just 1 wave for which there are systematically missing values or no answer, question E069_17 for wave 4, which was not asked (-4). Apart from that, there are no other questions for which there is systematically no answer, which is positive in terms of the amount of information that we have. 

## Save dataset with relevant questions to csv

In [50]:
df_us_concat_subset.to_csv('US_WVS_timeseries_relevant.csv', index = False)

## Clean dataset with non-relevant answers and aggregating by survey wave using weights
Remember the interepretation of weights: https://www.worldvaluessurvey.org/WVSContents.jsp?CMSID=WEIGHT&CMSID=WEIGHT. Below, we will use the weight `S017`, which is a weighting method based on "each participant [of a] country" (instead of `S018` which is intended for cross-country comparisons).

In [44]:
df = pd.read_csv('US_WVS_timeseries_relevant.csv')

df.head()

S002VS  S003        S007     S017      S018  S020  C006  E069_06  E069_07  \
0       3   840  8400320001  0.50021  0.648508  1995     9        2        2   
1       3   840  8400320002  0.87037  0.648508  1995     8        1        3   
2       3   840  8400320003  1.42561  0.648508  1995     2        3        3   
3       3   840  8400320004  0.88037  0.648508  1995     9        2        2   
4       3   840  8400320005  0.65028  0.648508  1995     2        2        3   

   E069_17  G006  A009  
0        2     1     2  
1        3     1     1  
2        3     2     3  
3        2     1     1  
4        3     2     1

Legend for the variables:
1. **Work and Economy**: C006: How satisfied are you with the financial situation of your household? (from 1 - dissatisfied to 10 - satisfied).
2. **Attitude towards police**: E069_06: Confidence towards the police (from 1 - extreme confidence to 4 - no confidence at all).
3. **Politics and democracy**: E069_07: Confidence towards the parliament; E069_17: Confidence towards the justice system/courts (from 1 - extreme confidence to 4 - no confidence at all).
4. **Pride of nationality**: G006. How proud are you to be of nationality of this country? (from 1 - very proud to 4 - not proud at all).
5. **Health**:  A009: State of health (subjective) (from 1 - very good to 5 - very poor).

Now, we will invert the values for the all of the columns of interest except for `C006` (since the interpretation of those is antiintuitive - lower values are better) and normalize *all* the columns to [0, 1], according to the scale of the questions of the survey. In all cases, we ignore negative values (which will be converted to NaN), which denote that answers have not been given or that the question was not asked.

In [46]:
for column in ['E069_06', 'E069_07', 'E069_17', 'G006']:  # Columns with scale of 1 (max) to 4 (min)
    max_value = 4
    min_value = 1
    df[column] = np.where(df[column] >= 0, (max_value - df[column]) / (max_value - min_value), np.nan)  # Inverse normalization

for column in ['A009']:  # Columns with scale of 1 (max) to 5 (min)
    max_value = 5
    min_value = 1
    df[column] = np.where(df[column] >= 0, (max_value - df[column]) / (max_value - min_value), np.nan)  # Inverse normalization

for column in ['C006']:  # Columns that only have to be normalized, with scale of 1 (min) to 10 (max)
    max_value = 10
    min_value = 1
    df[column] = np.where(df[column] >= 0, (df[column] - min_value) / (max_value - min_value), np.nan)  # Regular normalization

df.head()

S002VS  S003        S007     S017      S018  S020      C006   E069_06  \
0       3   840  8400320001  0.50021  0.648508  1995  0.888889  0.666667   
1       3   840  8400320002  0.87037  0.648508  1995  0.777778  1.000000   
2       3   840  8400320003  1.42561  0.648508  1995  0.111111  0.333333   
3       3   840  8400320004  0.88037  0.648508  1995  0.888889  0.666667   
4       3   840  8400320005  0.65028  0.648508  1995  0.111111  0.666667   

    E069_07   E069_17      G006  A009  
0  0.666667  0.666667  1.000000  0.75  
1  0.333333  0.333333  1.000000  1.00  
2  0.333333  0.333333  0.666667  0.50  
3  0.666667  0.666667  1.000000  1.00  
4  0.333333  0.333333  0.666667  1.00

In [40]:
# Define columns and their respective scales
columns_info = {
    'E069_06': {'scale': (1, 4), 'normalize': 'inverse'},  # Column with scale of 1 (max) to 4 (min)
    'E069_07': {'scale': (1, 4), 'normalize': 'inverse'},  # Column with scale of 1 (max) to 4 (min)
    'E069_17': {'scale': (1, 4), 'normalize': 'inverse'},  # Column with scale of 1 (max) to 4 (min)
    'G006': {'scale': (1, 4), 'normalize': 'inverse'},  # Column with scale of 1 (max) to 4 (min)
    'A009': {'scale': (1, 5), 'normalize': 'inverse'},  # Column with scale of 1 (max) to 5 (min)
    'C006': {'scale': (1, 10), 'normalize': 'regular'}  # Column that only have to be normalized, with scale of 1 (min) to 10 (max)
}

# Function to normalize a column
def normalize_column(df, column, scale, method='regular'):
    max_value, min_value = scale
    if method == 'inverse':
        # Inverse normalization: higher values become lower and vice versa, ignoring negative values
        df[column] = np.where(df[column] >= 0, (max_value - df[column]) / (max_value - min_value), np.nan)
    elif method == 'regular':
        # Regular normalization: scale between 0 and 1
        df[column] = np.where(df[column] >= 0, (df[column] - min_value) / (max_value - min_value), np.nan)
    return df

# Apply normalization to each column
for column, info in columns_info.items():
    df = normalize_column(df, column, info['scale'], info['normalize'])

# Display the first few rows of the modified DataFrame
df.head()

S002VS  S003        S007     S017      S018  S020      C006   E069_06  \
0       3   840  8400320001  0.50021  0.648508  1995  1.012346 -0.111111   
1       3   840  8400320002  0.87037  0.648508  1995  1.024691 -0.000000   
2       3   840  8400320003  1.42561  0.648508  1995  1.098765 -0.222222   
3       3   840  8400320004  0.88037  0.648508  1995  1.012346 -0.111111   
4       3   840  8400320005  0.65028  0.648508  1995  1.098765 -0.111111   

    E069_07   E069_17      G006    A009  
0 -0.111111 -0.111111 -0.000000 -0.0625  
1 -0.222222 -0.222222 -0.000000 -0.0000  
2 -0.222222 -0.222222 -0.111111 -0.1250  
3 -0.111111 -0.111111 -0.000000 -0.0000  
4 -0.222222 -0.222222 -0.111111 -0.0000

Now, the legend for the variables is all normalized to [0, 1], with 1 representing the best of each answer and 0 the worst.
1. **Work and Economy**: C006: How satisfied are you with the financial situation of your household? 
2. **Attitude towards police**: E069_06: Confidence towards the police 
3. **Politics and democracy**: E069_07: Confidence towards the parliament; E069_17: Confidence towards the justice system/courts 
4. **Pride of nationality**: G006. How proud are you to be of nationality of this country?
5. **Health**:  A009: State of health (subjective)

In [47]:
# Define values to remove
# values_to_remove = {-5, -4, -3, -2, -1}  # Original values
values_to_remove = {np.nan}  # Values were converted to NaN!

columns_of_interest = ['C006', 'E069_06', 'E069_07', 'E069_17', 'G006', 'A009']

# Get unique survey waves
survey_waves = sorted(df['S002VS'].unique())

# Initialize avg_values_wave with NaN for each column and wave
avg_values_wave = {'S002VS': survey_waves}
for column in columns_of_interest:
    avg_values_wave[column] = [np.nan] * len(survey_waves)

def weighted_avg(group, value_column, weight_column):  # Function to compute weighted average
    v = group[value_column]
    w = group[weight_column]
    return (v * w).sum() / w.sum()

# Loop over all columns of interest
for column in columns_of_interest:
    # Filter out rows containing the values we want to remove
    df_filtered = df[~df[column].isin(values_to_remove)]
    # Calculate the weighted average of each column per wave using 'S017' as weights
    weighted_average_column = df_filtered.groupby('S002VS', dropna=False).apply(
        weighted_avg, column, 'S017', include_groups = False).reset_index(name=f'Weighted_Avg_{column}')
    # Merge the weighted averages with the avg_values_wave dictionary
    for i, wave in enumerate(survey_waves):
        # Find the weighted average for the current wave
        avg_value = weighted_average_column.loc[weighted_average_column['S002VS'] == wave, f'Weighted_Avg_{column}']
        # If the wave is present, update the avg_values_wave dictionary
        if not avg_value.empty:
            avg_values_wave[column][i] = avg_value.values[0]

# Convert the dictionary to a DataFrame
df_avg_values = pd.DataFrame(avg_values_wave)

df_avg_values

S002VS      C006   E069_06   E069_07   E069_17      G006      A009
0       1  0.632990  0.670595  0.533722  0.544818  0.917621  0.769726
1       2  0.657349  0.643855  0.472726  0.538378  0.913579  0.782325
2       3  0.608330  0.608075  0.397362  0.429487  0.919546  0.788468
3       4  0.614525  0.613653  0.439255       NaN  0.891070  0.807897
4       5  0.539575  0.610382  0.358577  0.535313  0.854406  0.768727
5       6  0.572030  0.601316  0.341008  0.524487  0.830253  0.762933
6       7  0.579126  0.610299  0.309971  0.541079  0.744109  0.720816

In [50]:
# Add columnn with years of survey wave
period_mapping = {
    1: '1981-1984',
    2: '1989-1993',
    3: '1994-1998',
    4: '1999-2004',
    5: '2005-2009',
    6: '2010-2014',
    7: '2017-2022'
}

df_avg_values['period'] = df_avg_values['S002VS'].map(period_mapping)

df_avg_values

S002VS      C006   E069_06   E069_07   E069_17      G006      A009  \
0       1  0.632990  0.670595  0.533722  0.544818  0.917621  0.769726   
1       2  0.657349  0.643855  0.472726  0.538378  0.913579  0.782325   
2       3  0.608330  0.608075  0.397362  0.429487  0.919546  0.788468   
3       4  0.614525  0.613653  0.439255       NaN  0.891070  0.807897   
4       5  0.539575  0.610382  0.358577  0.535313  0.854406  0.768727   
5       6  0.572030  0.601316  0.341008  0.524487  0.830253  0.762933   
6       7  0.579126  0.610299  0.309971  0.541079  0.744109  0.720816   

      period  
0  1981-1984  
1  1989-1993  
2  1994-1998  
3  1999-2004  
4  2005-2009  
5  2010-2014  
6  2017-2022

Note how the value is missing for wave 4, variable E069_17 (for which this question was not asked)!

In [59]:
# Rename variables for clarity
rename_map = {
    'S002VS': 'survey_wave_number',
    'C006': 'financial_situation', 
    'E069_06': 'confidence_police', 
    'E069_07': 'confidence_parliament', 
    'E069_17': 'confidence_courts',
    'G006': 'pride_nationality',
    'A009': 'state_health'
}

df_avg_values = df_avg_values.rename(columns = rename_map)

df_avg_values

survey_wave_number  financial_situation  confidence_police  \
0                   1             0.632990           0.670595   
1                   2             0.657349           0.643855   
2                   3             0.608330           0.608075   
3                   4             0.614525           0.613653   
4                   5             0.539575           0.610382   
5                   6             0.572030           0.601316   
6                   7             0.579126           0.610299   

   confidence_parliament  confidence_courts  pride_nationality  state_health  \
0               0.533722           0.544818           0.917621      0.769726   
1               0.472726           0.538378           0.913579      0.782325   
2               0.397362           0.429487           0.919546      0.788468   
3               0.439255                NaN           0.891070      0.807897   
4               0.358577           0.535313           0.854406      0.768727   
5               0.341008           0.524487           0.830253      0.762933   
6               0.309971           0.541079           0.744109      0.720816   

      period  
0  1981-1984  
1  1989-1993  
2  1994-1998  
3  1999-2004  
4  2005-2009  
5  2010-2014  
6  2017-2022

In [61]:
df_avg_values.to_csv('US_WVS_timeseries_aggregated.csv', index = False)

In [2]:
df_avg_values = pd.read_csv('US_WVS_timeseries_aggregated.csv')

In [5]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

columns_of_interest = [col for col in df_avg_values.columns if col not in ['survey_wave_number', 'period']]

# Determine the number of columns for the subplot grid
num_columns = 3
num_rows = (len(columns_of_interest) + num_columns - 1) // num_columns

# Create a subplot figure
fig = make_subplots(
    rows=num_rows, cols=num_columns,
    shared_xaxes=True,
    subplot_titles=columns_of_interest
)

# Add a line plot for each column of interest
for i, column in enumerate(columns_of_interest):
    row = i // num_columns + 1
    col = i % num_columns + 1
    fig.add_trace(
        go.Scatter(
            x=df_avg_values['period'],
            y=df_avg_values[column],
            mode='lines+markers',
            name=column
        ),
        row=row, col=col
    )

# Update layout
fig.update_layout(
    height=400 * num_rows,  # Adjust height as needed
    width=1500,
    title_text='Weighted Average Values by Survey Wave',
    showlegend=False  # Hide legend to avoid redundancy
)

# Show the plot
fig.show()